README.md
# TYPO3 Extension hosts_pattern

[![Latest Stable Version](https://img.shields.io/packagist/v/ichhabrecht/hosts-pattern.svg)](https://packagist.org/packages/ichhabrecht/hosts-pattern)
[![Build Status](https://img.shields.io/travis/IchHabRecht/hosts_pattern/master.svg)](https://travis-ci.org/IchHabRecht/hosts_pattern)
[![StyleCI](https://styleci.io/repos/20126776/shield?branch=master)](https://styleci.io/repos/20126776)

Set TYPO3_CONF_VARS['SYS']['trustedHostsPattern'] depending on available sys_domain records

## Installation

Simply install the extension with Composer or the Extension Manager.

## Usage

1. Open `Hosts Pattern` backend module

2. Check domain records that were found and pattern that was built

3. Press `Write configuration` button to write the built pattern to your LocalConfiguration.php file


ChangeLog
2018-10-03         [RELEASE] Release of hosts_pattern 2.3.3 (Nicole Cordes)
2018-10-03 a385681 [TASK] Raise TYPO3 dependencies to include 9.5 (Nicole Cordes)
2018-10-03 aa0bf53 [TASK] Update supporting files to EXT:my_skeleton 0.1.5 (Nicole Cordes)
2018-10-03 50298de [BUGFIX] Bound module to system maintainers (Nicole Cordes)

2018-04-15         [RELEASE] Release of hosts_pattern 2.3.2 (Nicole Cordes)
2018-04-15 929d050 [TASK] Add additional typo3/cms composer configuration (Nicole Cordes)
2018-04-15 1be0c47 [TASK] Raise TYPO3 dependencies to include 9.2 (Nicole Cordes)
2018-02-01 798fd50 [TASK] Clean up TYPO3 dependencies (Nicole Cordes)
2018-01-31 df95967 [TASK] Update composer.json license definition (Nicole Cordes)
2018-01-31 b08d306 [TASK] Prefer stable versions for dev-branches on Travis CI tests (Nicole Cordes)

2018-01-31         [RELEASE] Release of hosts_pattern 2.3.1 (Nicole Cordes)
2018-01-31 969da4c [TASK] Raise TYPO3 dependencies to include 9.1 (Nicole Cordes)

2017-12-21         [RELEASE] Release of hosts_pattern 2.3.0 (Nicole Cordes)
2017-12-21 5df2ae9 [TASK] Add TYPO3 9.0 to extension constraints and Travis CI (Nicole Cordes)

2017-12-21         [RELEASE] Release of hosts_pattern 2.2.0 (Nicole Cordes)
2017-12-21 83c28a7 [FEATURE] Add new stage to automatically publish to TER (Nicole Cordes)
2017-12-21 186e4ad [TASK] Add README.md (Nicole Cordes)
2017-12-21 e4baf43 [BUGFIX] Fix code smell according to SonarQube Scanner (Nicole Cordes)
2017-12-21 f64a437 [FEATURE] Add SonarQube Scanner to Travis CI (Nicole Cordes)
2017-12-20 1eec9ba [FEATURE] Add stages to Travis CI configuration (Nicole Cordes)
2017-12-20 adf2dee [TASK] Update PHP-CS-Fixer and StyleCI configuration (Nicole Cordes)
2017-12-20 624f613 [TASK] Update to latest nimut/testing-framework (Nicole Cordes)
2017-12-20 2745189 [BUGFIX] Provide optimized module SVG icon (Nicole Cordes)
2017-12-20 1326d2c [TASK] Raise compatibility to TYPO3 8.7 (Nicole Cordes)
2017-03-05 3912ad4 [TASK] Add more coding style rules (Nicole Cordes)
2017-03-03 351ff39 [TASk] Move friendsofphp/php-cs-fixer to suggest section (Nicole Cordes)
2017-03-03 c132ed0 [TASK] Raise nimut/testing-framework dependency to 1.x-dev (Nicole Cordes)
2017-02-23 c15be2e [TASK] Raise nimut/testing-framework dependency to 1.0 (Nicole Cordes)
2017-02-21 1e68cf9 [TASK] Fix coding style according to php-cs-fixer (Nicole Cordes)
2017-02-21 4a4049d [FEATURE] Add StyleCI Coding Style Service (Nicole Cordes)
2017-02-21 ad83946 [FEATURE] Add friendsofphp/php-cs-fixer as dev requirement (Nicole Cordes)
2017-02-21 40018ca [TASK] Exclude local .php_cs files in .gitignore (Nicole Cordes)

2017-02-20         [RELEASE] Release of hosts_pattern 2.1.3 (Nicole Cordes)
2017-02-20 8e14c28 [TASK] Raise compatibility to TYPO3 8.6 (Nicole Cordes)
2017-02-20 1a13bdb [TASK] Raise nimut/testing-framework to ^0.2 (Nicole Cordes)
2017-02-16 c8ce0df [TASK] Use test class from nimut/testing-framework (Nicole Cordes)
2017-02-16 2474e65 [TASK] Use nimut/testing-framework for unit tests (Nicole Cordes)

2016-12-22         [RELEASE] Release of hosts_pattern 2.1.2 (Nicole Cordes)
2016-12-22 f9cdb57 [TASK] Raise compatibility to TYPO3 8.5 (Nicole Cordes)
2016-11-08 b6c44cd [TASK] Add more matrix versions to .travis.yml (Nicole Cordes)

2016-11-08         [RELEASE] Release of hosts_pattern 2.1.1 (Nicole Cordes)
2016-11-08 42a891b [TASK] Raise version compatibility to TYPO3 8.4.99 (Nicole Cordes)

2016-08-10         [RELEASE] Release of hosts_pattern 2.1.0 (Nicole Cordes)
2016-08-10 3cd9196 [TASK] Adjust author email address (Nicole Cordes)
2016-08-10 0ea43f9 [TASK] Raise TYPO3 version to 8.2.99 (Nicole Cordes)
2016-08-05 e216686 [TASK] Add travis.yml (Nicole Cordes)
2016-08-05 be8f4e1 [TASK] Ignore .Build folder (Nicole Cordes)
2016-08-05 0178c6d [TASK] Add composer.json (Nicole Cordes)
2016-08-05 aceb1e8 [BUGFIX] Prevent ouput escaping in IconViewHelper (Nicole Cordes)
2016-08-05 926de52 [BUGFIX] Add missing use statement for Domain model (Nicole Cordes)
2016-08-05 b014330 [TASK] Add new extension icon (Nicole Cordes)

2015-10-29         [RELEASE] Release of hosts_pattern 2.0.0 (Nicole Cordes)
2015-10-29 b2b4a74 [TASK] Reformat code according to TYPO3 Coding Guidelines (Nicole Cordes)
2015-10-29 e8ce704 [BUGFIX] Use new module api (Nicole Cordes)
2015-10-29 97a8527 [BUGFIX] Adjust iconViewHelper to new icon api (Nicole Cordes)
2015-10-29 2aa538a [TASK] Add namespaces to extension (Nicole Cordes)
2015-10-29 9edf533 [TASK] Set dependency to TYPO3 7.5 (Nicole Cordes)
2015-10-29 b0f380f [TASK] Replace deprecated class Tx_Extbase_Persistence_QuerySettingsInterface (Nicole Cordes)
2015-10-29 4538194 [TASK] Replace deprecated class Tx_Extbase_Persistence_Repository (Nicole Cordes)
2015-10-29 2907d22 [TASK] Replace deprecated class Tx_Extbase_DomainObject_AbstractEntity (Nicole Cordes)
2015-10-29 e8ca767 [TASK] Replace deprecated class Tx_Extbase_MVC_Controller_ActionController (Nicole Cordes)
2015-10-29 3e2b263 [TASK] Replace deprecated class t3lib_div (Nicole Cordes)
2015-10-29 c3914e8 [TASK] Remove legacy code for minor TYPO3 versions (Nicole Cordes)
2015-10-29 83cfee1 [TASK] Replace deprecated class t3lib_extMgm (Nicole Cordes)
2015-10-29 4c8319c [TASK] Replace deprecated class Tx_Extbase_Utility_Extension (Nicole Cordes)

2015-10-29         [RELEASE] Release of hosts_pattern 1.3.0 (Nicole Cordes)
2015-02-18 cee4061 [BUGFIX] Use old flashMessageContainer for TYPO3 6.1 (Nicole Cordes)

2014-09-16         [RELEASE] Release of hosts_pattern 1.2.2 (Nicole Cordes)
2014-09-16 180e1a9 [HOTFIX] Reverse array before implode to reorder subdomains (Nicole Cordes)

2014-08-26         [RELEASE] Release of hosts_pattern 1.2.1 (Nicole Cordes)
2014-08-26 acc20a6 [HOTFIX] Add support for PHP 5.2 (Nicole Cordes)

2014-05-25         [RELEASE] Release of hosts_pattern 1.2.0 (Nicole Cordes)
2014-05-25 26cf045 [TASK] Improve unit tests (Nicole Cordes)

Tests/Unit/Service/PatternServiceTest.php
<?php
namespace IchHabRecht\HostsPattern\Tests\Unit;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use IchHabRecht\HostsPattern\Domain\Model\Domain;
use IchHabRecht\HostsPattern\Service\PatternService;
use Nimut\TestingFramework\TestCase\UnitTestCase;

/**
 * Test case for IchHabRecht\HostsPattern\Service\PatternService
 */
class PatternServiceTest extends UnitTestCase
{
    /**
     * @var PatternService
     */
    protected $fixture;

    /**
     * @return void
     */
    public function setUp()
    {
        $this->fixture = new PatternService();
    }

    /**
     * @param array $domainArray
     * @return array
     */
    protected function getDomainObjectsFromArray(array $domainArray)
    {
        $domainStorageArray = [];
        foreach ($domainArray as $domain) {
            $domainObject = new Domain();
            $domainObject->setDomainName($domain);
            $domainStorageArray[] = $domainObject;
        }
        unset($domain);

        return $domainStorageArray;
    }

    /**
     * @return array
     */
    public function generatePatternDataProvider()
    {
        return [
            'Simple domain' => [
                [
                    'example.com',
                ],
                'example\\.com',
            ],
            'Same domain with different subdomains' => [
                [
                    'www.example.com',
                    'wwwt.example.com',
                ],
                '(www|wwwt)\\.example\\.com',
            ],
            'Same domain without subdomain' => [
                [
                    'example.com',
                    'www.example.com',
                ],
                '(www\\.)?example\\.com',
            ],
            'Same domain with multiple subdomains' => [
                [
                    'www.example.com',
                    'www.secure.example.com',
                ],
                '(www|www\\.secure)\\.example\\.com',
            ],
            'Two different domains' => [
                [
                    'www.example.com',
                    'www.domain.com',
                ],
                '(www\\.example\\.com|www\\.domain\\.com)',
            ],
            'Two equal domains with extra domain' => [
                [
                    'www.example.com',
                    'subdomain.example.com',
                    'www.domain.com',
                ],
                '((www|subdomain)\\.example\\.com|www\\.domain\\.com)',
            ],
            'Three equal domains without subdomain with extra domain' => [
                [
                    'www.example.com',
                    'subdomain.example.com',
                    'example.com',
                    'www.domain.com',
                ],
                '(((www|subdomain)\\.)?example\\.com|www\\.domain\\.com)',
            ],
        ];
    }

    /**
     * @test
     * @dataProvider generatePatternDataProvider
     *
     * @param array $domainArray
     * @param array $expectedPattern
     */
    public function generatePatternReturnsPatternForGivenDomains($domainArray, $expectedPattern)
    {
        $this->assertSame($expectedPattern,
            $this->fixture->generatePattern($this->getDomainObjectsFromArray($domainArray)));
    }

    /**
     * @test
     */
    public function generatePatternReturnsHttpHostWithoutDomain()
    {
        $_SERVER['HTTP_HOST'] = 'www.foo.bar';

        $this->assertSame('www\\.foo\\.bar', $this->fixture->generatePattern([]));
    }
}


ext_tables.php
<?php
if (!defined('TYPO3_MODE')) {
    die('Access denied.');
}

if (TYPO3_MODE === 'BE') {
    \TYPO3\CMS\Extbase\Utility\ExtensionUtility::registerModule(
        'IchHabRecht.hosts_pattern',
        'tools',
        'HostsPattern',
        '',
        [
            'Domain' => 'index',
        ],
        [
            'access' => version_compare(TYPO3_version, '9', '>=') ? 'systemMaintainer' : 'admin',
            'icon' => 'EXT:hosts_pattern/Resources/Public/Icons/module-hostspattern.svg',
            'labels' => 'LLL:EXT:hosts_pattern/Resources/Private/Language/locallang_mod.xml',
        ]
    );

    \TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addTypoScript(
        'hosts_pattern',
        'setup',
        '<INCLUDE_TYPOSCRIPT: source="FILE:EXT:hosts_pattern/Configuration/TypoScript/setup.txt">'
    );
}


Classes/Domain/Model/Domain.php
<?php
namespace IchHabRecht\HostsPattern\Domain\Model;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;

/**
 * Domain model
 */
class Domain extends AbstractEntity
{
    /**
     * @var string
     */
    protected $domainName;

    /**
     * @return string
     */
    public function getDomainName()
    {
        return $this->domainName;
    }

    /**
     * @param string $domainName
     * @return void
     */
    public function setDomainName($domainName)
    {
        $this->domainName = $domainName;
    }
}


Migrations/Code/ClassAliasMap.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

return [
    'Tx_Extbase_Persistence_QuerySettings' => 'TYPO3\\CMS\\Extbase\\Persistence\\Generic\\Typo3QuerySettings',
];


Classes/Service/PatternService.php
<?php
namespace IchHabRecht\HostsPattern\Service;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use IchHabRecht\HostsPattern\Domain\Model\Domain;
use TYPO3\CMS\Core\Utility\GeneralUtility;

/**
 * Service to generate regular expression pattern for trustedHostsPattern
 */
class PatternService
{
    /**
     * @param Domain[] $domainArray
     * @return string
     */
    public function generatePattern($domainArray)
    {
        if (empty($domainArray)) {
            $domain = new Domain();
            $domain->setDomainName(GeneralUtility::getIndpEnv('HTTP_HOST'));
            $domainArray = [$domain];
        }

        $patternArray = [];
        $extractedDomainsArray = $this->extractDomains($domainArray);
        foreach ($extractedDomainsArray as $domainName => $subdomainArray) {
            $patternArray[] = $this->getPatternForDomain($domainName, $subdomainArray);
        }
        unset($domain);

        return count($patternArray) > 1 ? '(' . implode('|', $patternArray) . ')' : $patternArray[0];
    }

    /**
     * @param Domain[] $domainArray
     * @return array
     */
    protected function extractDomains($domainArray)
    {
        $extractedDomainArray = [];

        /** @var Domain $domain */
        foreach ($domainArray as $domain) {
            $reverseHost = array_reverse(explode('.', $domain->getDomainName()));
            $domainName = preg_quote($reverseHost[1] . '.' . $reverseHost[0]);
            unset($reverseHost[0], $reverseHost[1]);
            $subdomain = '';
            if (!empty($reverseHost)) {
                $subdomain = implode('.', array_reverse($reverseHost));
            }
            if (!is_array($extractedDomainArray[$domainName])) {
                $extractedDomainArray[$domainName] = [];
            }
            $extractedDomainArray[$domainName][] = preg_quote($subdomain);
        }
        unset($domain);

        return $extractedDomainArray;
    }

    /**
     * @param string $domainName
     * @param array $subdomainArray
     * @return string
     */
    protected function getPatternForDomain($domainName, $subdomainArray)
    {
        $hasDomainWithoutSubdomain = false;
        if (in_array('', $subdomainArray)) {
            $hasDomainWithoutSubdomain = true;
            $subdomainArray = array_filter($subdomainArray, 'strlen');
        }
        $hasMultipleSubdomains = count($subdomainArray) > 1;
        $pattern = ($hasMultipleSubdomains ? '(' : '') . implode('|',
                $subdomainArray) . ($hasMultipleSubdomains ? ')' : '');
        if (!empty($subdomainArray)) {
            $pattern .= '\\.';
        }
        if ($hasDomainWithoutSubdomain && !empty($subdomainArray)) {
            $pattern = '(' . $pattern . ')?';
        }
        $pattern .= $domainName;

        return $pattern;
    }
}


Classes/Controller/DomainController.php
<?php
namespace IchHabRecht\HostsPattern\Controller;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use IchHabRecht\HostsPattern\Domain\Model\Domain;
use IchHabRecht\HostsPattern\Domain\Repository\DomainRepository;
use IchHabRecht\HostsPattern\Service\PatternService;
use TYPO3\CMS\Backend\View\BackendTemplateView;
use TYPO3\CMS\Core\Configuration\ConfigurationManager;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Extbase\Mvc\Controller\ActionController;

/**
 * Controller for backend module
 */
class DomainController extends ActionController
{
    /**
     * @var DomainRepository;
     */
    protected $domainRepository;

    /**
     * @var PatternService
     */
    protected $patternService;

    /**
     * View object name
     *
     * @var string
     */
    protected $defaultViewObjectName = BackendTemplateView::class;

    /**
     * @param DomainRepository $domainRepository
     */
    public function injectDomainRepository(DomainRepository $domainRepository)
    {
        $this->domainRepository = $domainRepository;
    }

    /**
     * @param PatternService $patternService
     * @return void
     */
    public function injectPatternService(PatternService $patternService)
    {
        $this->patternService = $patternService;
    }

    /**
     * @return void
     */
    public function indexAction()
    {
        $domains = $this->domainRepository->findAll();
        if (!count($domains)) {
            $domain = $this->objectManager->get(Domain::class);
            $domain->setDomainName(GeneralUtility::getIndpEnv('HTTP_HOST'));
            $domains = [$domain];
        }
        $pattern = $this->patternService->generatePattern($domains);

        if ($this->request->hasArgument('write')) {
            /** @var ConfigurationManager $configurationManager */
            $configurationManager = $this->objectManager->get(ConfigurationManager::class);
            $configurationManager->setLocalConfigurationValueByPath('SYS/trustedHostsPattern', $pattern);
            $this->addFlashMessage(
                htmlspecialchars('$GLOBALS[TYPO3_CONF_VARS][SYS][trustedHostsPattern] = ' . $pattern),
                'Trusted Hosts Pattern was changed'
            );
        }

        $this->view->assignMultiple([
            'domains' => $domains,
            'pattern' => $pattern,
        ]);
    }
}


Classes/Domain/Repository/DomainRepository.php
<?php
namespace IchHabRecht\HostsPattern\Domain\Repository;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use TYPO3\CMS\Extbase\Persistence\Generic\QuerySettingsInterface;
use TYPO3\CMS\Extbase\Persistence\Repository;

/**
 * Repository for domain records
 */
class DomainRepository extends Repository
{
    /**
     * @return void
     */
    public function initializeObject()
    {
        /** @var QuerySettingsInterface $defaultQuerySettings */
        $defaultQuerySettings = $this->objectManager->get(QuerySettingsInterface::class);
        $defaultQuerySettings->setRespectStoragePage(false);
        $this->setDefaultQuerySettings($defaultQuerySettings);
    }
}


Classes/ViewHelpers/Be/Buttons/IconViewHelper.php
<?php
namespace IchHabRecht\HostsPattern\ViewHelpers\Be\Buttons;

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2014 Nicole Cordes <cordes@cps-it.de>, CPS-IT GmbH
 *
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

use TYPO3\CMS\Core\Imaging\Icon;
use TYPO3\CMS\Core\Imaging\IconFactory;
use TYPO3\CMS\Core\Type\Icon\IconState;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Fluid\Core\Rendering\RenderingContextInterface;
use TYPO3\CMS\Fluid\Core\ViewHelper\AbstractViewHelper;
use TYPO3\CMS\Fluid\Core\ViewHelper\Facets\CompilableInterface;

/**
 * Class Tx_HostsPattern_ViewHelpers_Be_Buttons_IconViewHelper to render sprite icons
 */
class IconViewHelper extends AbstractViewHelper implements CompilableInterface
{
    /**
     * View helper returns HTML, thus we need to disable output escaping
     *
     * @var bool
     */
    protected $escapeOutput = false;

    public function initializeArguments()
    {
        parent::initializeArguments();
        $this->registerArgument('uri', 'string', '', false, '');
        $this->registerArgument('identifier', 'string', '', false, 'closedok');
        $this->registerArgument('title', 'string', '', false, '');
        $this->registerArgument('size', 'string', '', false, Icon::SIZE_SMALL);
        $this->registerArgument('overlay', 'string', '');
        $this->registerArgument('state', 'string', '', false, IconState::STATE_DEFAULT);
    }

    /**
     * Renders an sprite icon
     *
     * @return string
     */
    public function render()
    {
        return static::renderStatic(
            $this->arguments,
            $this->buildRenderChildrenClosure(),
            $this->renderingContext
        );
    }

    /**
     * @param array $arguments
     * @param \Closure $renderChildrenClosure
     * @param RenderingContextInterface $renderingContext
     * @return string
     */
    public static function renderStatic(array $arguments, \Closure $renderChildrenClosure, RenderingContextInterface $renderingContext)
    {
        $identifier = $arguments['identifier'];
        $size = $arguments['size'];
        $overlay = $arguments['overlay'];
        $state = IconState::cast($arguments['state']);

        /** @var IconFactory $iconFactory */
        $iconFactory = GeneralUtility::makeInstance(IconFactory::class);
        $icon = $iconFactory->getIcon($identifier, $size, $overlay, $state)->render();

        $title = $arguments['title'];
        if (!empty($title)) {
            $icon = '<span title="' . htmlspecialchars($title) . '">' . $icon . '</span>';
        }

        $uri = $arguments['uri'];
        if (empty($uri)) {
            return $icon;
        }

        return '<a href="' . htmlspecialchars($uri) . '">' . $icon . '</a>';
    }
}


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "hosts_pattern".
 *
 * Auto generated 03-10-2018 19:50
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
  'title' => 'Trusted Hosts Pattern',
  'description' => 'Set TYPO3_CONF_VARS[\'SYS\'][\'trustedHostsPattern\'] depending on available sys_domain records',
  'category' => 'module',
  'author' => 'Nicole Cordes',
  'author_email' => 'typo3@cordes.co',
  'author_company' => 'biz-design',
  'state' => 'stable',
  'uploadfolder' => 0,
  'createDirs' => '',
  'clearCacheOnLoad' => 0,
  'version' => '2.3.3',
  'constraints' => 
  array (
    'depends' => 
    array (
      'typo3' => '7.6.0-9.5.99',
    ),
    'conflicts' => 
    array (
    ),
    'suggests' => 
    array (
    ),
  ),
  '_md5_values_when_last_written' => 'a:21:{s:9:"ChangeLog";s:4:"04cb";s:7:"LICENSE";s:4:"b234";s:9:"README.md";s:4:"0315";s:13:"composer.json";s:4:"80b9";s:12:"ext_icon.png";s:4:"dc44";s:14:"ext_tables.php";s:4:"589d";s:16:"phpunit.xml.dist";s:4:"7d0d";s:24:"sonar-project.properties";s:4:"c87e";s:39:"Classes/Controller/DomainController.php";s:4:"b8d5";s:31:"Classes/Domain/Model/Domain.php";s:4:"9191";s:46:"Classes/Domain/Repository/DomainRepository.php";s:4:"61d9";s:34:"Classes/Service/PatternService.php";s:4:"a6ba";s:49:"Classes/ViewHelpers/Be/Buttons/IconViewHelper.php";s:4:"4b78";s:34:"Configuration/TypoScript/setup.txt";s:4:"7901";s:33:"Migrations/Code/ClassAliasMap.php";s:4:"e1b8";s:44:"Resources/Private/Language/locallang_mod.xlf";s:4:"1b6e";s:44:"Resources/Private/Language/locallang_mod.xml";s:4:"58bc";s:38:"Resources/Private/Layouts/Default.html";s:4:"7847";s:45:"Resources/Private/Templates/Domain/Index.html";s:4:"1855";s:46:"Resources/Public/Icons/module-hostspattern.svg";s:4:"dc49";s:41:"Tests/Unit/Service/PatternServiceTest.php";s:4:"e085";}',
);

